### ISSUE TITLE
Warn when committing in headless state

### ISSUE THREAD

**Issue by dpjohnst** [sublimehq/sublime_merge#797]
Migrated issue from: https://forum.sublimetext.com/t/feature-request-raise-a-warning-before-allowing-the-user-to-commit-when-in-a-headless-state/51994

**Problem description**
I often find myself having checked out a commit (instead of a branch) and performing modifications and committing, only to find that the commit that I just did is not on a branch and will be lost if I checkout another branch/commit. 

**Preferred solution**

Add a warning when the user commits in a headless state.

**Alternatives**

Warn when a checkout action would result in dangling commits.

**Comment by BenjaminSchaaf** [comment]
I think the best approach would be to make being in a detached state more obvious. I've seen people commit from a detached head as part of their workflow before, so I don't think it makes sense to deviate from git here with a model popup or similar. If someone wants to completely prevent committing from a detached head, they can use a pre-commit hook.

**Comment by dpjohnst** [comment]
For those running into this issue, the following link contains a pre-commit hook which can be used to prevent committing with a detached head: https://stackoverflow.com/questions/37269048/how-to-prevent-commit-in-detached-head

(Credit to @TheSecEng for sharing on the forum thread)

**Comment by stinos** [comment]
Second the pre-commit hook: SM should not try to be opinionated and enforce or hinder any type of git workflow. There's nothing inherently wrong with commiting with detached HEAD.

**Comment by srbs** [comment]
See related: #342

**Comment by zukoj** [comment]
> Second the pre-commit hook: SM should not try to be opinionated and enforce or hinder any type of git workflow. There's nothing inherently wrong or different for commiting with detached HEAD.

I agree but there is a risk of losing data when commits are performed in detached HEAD. SM should protect the user against himself if possible.

**Comment by TheSecEng** [comment]
> > Second the pre-commit hook: SM should not try to be opinionated and enforce or hinder any type of git workflow. There's nothing inherently wrong or different for commiting with detached HEAD.
> 
> I agree but there is a risk of losing data when commits are performed in detached HEAD. SM should protect the user against himself if possible.

That makes it opinionated, I think this should be a api or git hook. Not core feature.

**Comment by stinos** [comment]
> SM should protect the user against himself if possible.

I understand you feel that way, but I'm not sure this actually aligns with the original goal of the developers of SM; I thought they just want SM to be as close to git as possible and never get in the way. So just like git doesn't protect the user - only the user's knowledge can - nor does SM.

That being said: the Discard All/Stage All buttons also require clicking twice so there actually is a form of protection. But something like deleting all files is a destructive operation, comitting is not. So in my view the issue starts off on the wrong foot already: if you need protection, it shouldn't be against comitting. It should be when switching to another branch.

**Comment by BenjaminSchaaf** [comment]
> But something like deleting all files is a destructive operation, comitting is not. So in my view the issue starts off on the wrong foot already: if you need protection, it shouldn't be against comitting. It should be when switching to another branch.

Just to add to this: Checking out another commit isn't a destructive operation. Whatever objects become unreferenced are still around and can be recovered. Only when git does garbage collection will an operation actually become destructive. Git does a gc from a number of operations, including `commit`, `fetch` and `merge`.

As a more general solution it doesn't make much sense to always have a confirmation for "destructive" operations, as the destructive nature of an operation is rather arbitrary in git. Even if you define destructive as any time you lose a reference to an object (ie. it becomes gc-able), you'd be adding confirmation dialogs to ammends, commit message edits, fetches/pulls, stash pops, etc.

**Comment by stinos** [comment]
Yes good point, thanks for adding; I'm never sure how many people know about reflog etc.

> As a more general solution it doesn't make much sense to always have a confirmation for "destructive" operations

Has this been considered for revising? Because the only 'really' irrecoverable destructive operations I think are Discard All and Delete All. Stage All on the other hand now also requires clicking twice, for consistency perhaps, but at least for me it's rather annoying at times. But this goes to much into the realm of personal preference perhaps.

**Comment by BenjaminSchaaf** [comment]
Those buttons have confirmations because they're easy to hit accidentally and make a lot of difficult to reverse changes. If you think that behaviour should change that discussion should happen in a separate issue.
